VSR: Move formatting into `replica.zig`

sentientwaffle opened the pull request:
- Move formatting (from superblock) into `replica.zig`.
- Rewrite WAL formatting to be async (previously used `os.write()`).

Previously the `MessagePool` was owned by the `MessageBus`, but this introduced a circular dependency: `SuperBlock` requires a `MessagePool`, `MessageBus` requires `cluster`, `cluster` is retrieved from the `SuperBlock`. I resolved this by moving `MessagePool` out and passing `MessageBus` a reference to the pool.

The `MessageBus` is required by the `Replica`, but must be initialized outside because we support different implementations (for now — see https://github.com/coilhq/tigerbeetle/issues/71).
But it must be initialized _after_ the `SuperBlock` is opened, since it requires the `cluster`.

sentientwaffle committed:
VSR: Move formatting into `replica.zig`

- Move formatting (from superblock) into `replica.zig`.
- Rewrite WAL formatting to be async (previously used `os.write()`).

Previously the `MessagePool` was owned by the `MessageBus`, but this introduced a circular dependency: `SuperBlock` requires a `MessagePool`, `MessageBus` requires `cluster`, `cluster` is retrieved from the `SuperBlock`. I resolved this by moving `MessagePool` out and passing `MessageBus` a reference to the pool.

The `MessageBus` is required by the `Replica`, but must be initialized outside because we support different implementations (for now — see https://github.com/coilhq/tigerbeetle/issues/71).
But it must be initialized _after_ the `SuperBlock` is opened, since it requires the `cluster`.

kprotty commented:
Nice resolution for the dependencies. `SuperBlock` could take in a `MessageBus` (it uses it primarily to get/unref messages in `SuperBlockClientTable`), but `MessagePool` was used instead as its easier to setup for testing than the former.

sentientwaffle commented:
Yeah, `MessagePool` is better for `SuperBlock` — `MessageBus` would cause a circular dependency because it requires the `cluster`. (Plus using the narrower dependency is preferable in general).

jorangreef commented on src/main.zig:
```diff
@@ -16,10 +16,12 @@ const Time = @import("time.zig").Time;
 
 const vsr = @import("vsr.zig");
 const Replica = vsr.Replica(StateMachine, MessageBus, Storage, Time);
+const ReplicaFormatter = vsr.ReplicaFormatter(Storage);
```
```suggestion
const ReplicaFormat = vsr.ReplicaFormatType(Storage);
```

For type constructors, we've started adding `Type` as a suffix, e.g. `StateMachineType`. This helps reduce the chance of shadowing at the call site. For example, you get `StateMachine` from a `StateMachineType` — this helped clean up the LSM quite nicely since we have so many generics there.

I would suggest `ReplicaFormat` as a name, describing the format type as a noun not as a verb, so it's a little broader, and also works if we extend `Format` to do things like inspection or diagnostics or update of a format, i.e. not only the task of formatting, but anything to do with working with a format.

jorangreef replied:
For example, we may soon need to be able to take in a format, and then rewrite the replica index. This could be necessary when recovering a replica from a backup, where we copy the data file, rewrite the replica-specific index and then run the recovery protocol.

sentientwaffle replied:
> This helps reduce the chance of shadowing at the call site.

This is a great idea; it also allows you to easily grep the code for either the type constructor or the type independently.

jorangreef commented on src/main.zig:
```diff
@@ -168,33 +158,39 @@ const Command = struct {
             &command.io,
+            &message_pool,
         );
+        errdefer message_bus.deinit();
```
Do we also want to call `deinit()` for `message_pool` and `replica` etc.?

sentientwaffle replied:
`MessagePool` doesn't have a `deinit()` function. (This surprised me; I plan to add one when I have time).

Also, `MessageBus`'s `deinit()` is currently a no-op.

I'll add the call `deinit()` for `replica`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -209,19 +216,26 @@ pub fn Replica(
             cluster: u32,
             replica_count: u8,
-            replica: u8,
+            replica_index: u8,
```
Nice change!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+
+        journal_write: Storage.Write = undefined,
+        journal_write_offset: u64 = 0,
+        journal_write_buffer: []u8,
```
We need this buffer type to be aligned to a sector, and also allocated according to alignment. Otherwise, we may get EINVAL from the kernel because alignment is required for Direct I/O.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+        storage: *Storage,
+        callback: ?fn (formatter: *Self) void = null,
+
+        journal_write: Storage.Write = undefined,
```
We could rename the `journal_` prefixes here to `wal_` since we use that also a bit further down.

jorangreef replied:
Since we're only doing writes in this context, we could also just do `wal_offset`, `wal_buffer`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+            assert(self.journal_write_offset <= config.journal_size_max);
+
+            const offset = self.journal_write_offset;
+            const buffer = self.journal_write_buffer;
```
Better to “shrink the scope” to reduce probability of misuse, and rather use these directly inline without taking const references, to avoid introducing a variable into the scope, and to avoid potentially copying something to the stack (in general, not a problem here).

If these are renamed to `wal_offset` and `wal_buffer` then that will also make inlining them easier.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+
+            const offset = self.journal_write_offset;
+            const buffer = self.journal_write_buffer;
+            const write_size = vsr.format_journal(self.cluster, offset, buffer);
```
We could also just do `size` here since the context everywhere is for writing.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+                self.storage.write_sectors(
+                    format_wal_sectors_callback,
+                    &self.journal_write,
+                    buffer[0..write_size],
```
What are our expectations of the contents of the buffer that we are writing here, that `format_journal()` will ensure everything in the buffer is ready for writing up to `size`?

sentientwaffle replied:
Yes. I'll add an assertion to check:

```zig

            const zeros = [_]u8{0} ** @sizeOf(Header);
            for (std.mem.bytesAsSlice(Header, self.wal_buffer[0..size])) |*header| {
                if (header.checksum == 0 and header.checksum_body == 0) {
                    // This is the (empty) body of a Prepare.
                    assert(std.mem.eql(u8, std.mem.asBytes(header), &zeros));
                } else {
                    assert(header.valid_checksum());
                    assert(header.command == .prepare or header.command == .reserved);
                    assert(header.op != 0 or header.operation == .root);
                }
            }
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5165,3 +5181,111 @@ pub fn Replica(
+        fn format_wal_sectors_callback(write: *Storage.Write) void {
+            const self = @fieldParentPtr(Self, "journal_write", write);
+            assert(self.callback != null);
+            assert(self.journal_write_offset == config.journal_size_max);
```
I think this could fail? For example, if we have to call `format_wal_sectors()` several times?

sentientwaffle replied:
Oops, that should have been `<=`.

jorangreef reviewed (changes requested):
We need to guarantee alignment for Direct I/O and then perhaps there's an assertion in here that would fail if we call `format_wal_sectors()` more than once.

sentientwaffle committed:
VSR: Align format buffer; fix assertions

- Sector-align the buffer used to format the WAL, to avoid `EINVAL`.
- Fix assertion typo.
- Add assertions.
- Rename types to new convention.
- Misc code review cleanup.
